sqlserver/datadog_checks/sqlserver/stored_procedures.py

```diff
@@ -0,0 +1,186 @@
+            rate_limit=1 / float(collection_interval),
+            job_name="procedure-metrics",
+            shutdown_callback=self._close_db_conn,
+        )
+        self.dm_exec_procedure_stats_row_limit = int(
+            check.procedure_metrics_config.get('dm_exec_procedure_stats_row_limit', 10000)
+        )
+        self._state = StatementMetrics()
+        self._conn_key_prefix = "dbm-procedures"
+        self._procedure_metrics_query = None
+        self._last_stats_query_time = None
+        self._max_procedure_metrics = check.procedure_metrics_config.get("max_procedures", 250)
+
+    def _close_db_conn(self):
+        pass
+
+    def _get_available_procedure_metrics_columns(self, cursor, all_expected_columns):
+        cursor.execute("select top 0 * from sys.dm_exec_procedure_stats")
+        all_columns = {i[0] for i in cursor.description}
+        available_columns = [c for c in all_expected_columns if c in all_columns]
+        missing_columns = set(all_expected_columns) - set(available_columns)
+        if missing_columns:
+            self.log.debug(
+                "missing the following expected query metrics columns from dm_exec_procedure_stats: %s", missing_columns
+            )
+        self.log.debug("found available sys.dm_exec_procedure_stats columns: %s", available_columns)
+        return available_columns
+
+    def _get_procedure_metrics_query_cached(self, cursor):
+        if self._procedure_metrics_query:
+            return self._procedure_metrics_query
+        available_columns = self._get_available_procedure_metrics_columns(cursor, SQL_SERVER_PROCEDURE_METRICS_COLUMNS)
+        self._procedure_metrics_query = PROCEDURE_METRICS_QUERY.format(
+            procedure_metrics_columns=', '.join(['sum({}) as {}'.format(c, c) for c in available_columns]),
+            limit=self.dm_exec_procedure_stats_row_limit,
+        )
+        return self._procedure_metrics_query
+
+    @tracked_method(agent_check_getter=agent_check_getter, track_result_length=True)
+    def _load_raw_procedure_metrics_rows(self, cursor):
+        self.log.debug("collecting sql server procedure metrics")
+        procedure_metrics_query = self._get_procedure_metrics_query_cached(cursor)
+        now = time.time()
+        self._last_stats_query_time = now
+        self.log.debug("Running query [%s]", procedure_metrics_query)
+        cursor.execute(procedure_metrics_query)
+        columns = [i[0] for i in cursor.description]
+        # construct row dicts manually as there's no DictCursor for pyodbc
+        rows = [dict(zip(columns, row)) for row in cursor.fetchall()]
+        return rows
+
+    def _collect_metrics_rows(self, cursor):
+        rows = self._load_raw_procedure_metrics_rows(cursor)
+        rows = [r for r in rows if r.get('procedure_name') and r.get('database_name')]
+        if not rows:
+            return []
+        metric_columns = [c for c in rows[0].keys() if c.startswith("total_") or c == 'execution_count']
```

Comment:
curious why we need the filter `if c.startswith("total_") or c == 'execution_count'` if all expected metrics columns are pre-defined in `SQL_SERVER_PROCEDURE_METRICS_COLUMNS`? I think `metric_columns = list(rows[0].keys())` should be enough.

Reply:
Some columns like `database, procedure_name, schema_name` are not metrics and don't need to be diffed. But I could make this more resilient by doing `c for c in rows[0].keys() if c in SQL_SERVER_PROCEDURE_METRICS_COLUMNS`